File: backend/models/cluster.py
"""
Cluster and GPU node models for distributed kernel execution.
"""
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from enum import Enum


class NodeStatus(str, Enum):
    """Status of a cluster node."""
    ONLINE = "online"
    OFFLINE = "offline"
    BUSY = "busy"
    ERROR = "error"
    MAINTENANCE = "maintenance"


class GPUInfo(BaseModel):
    """GPU information for a node."""
    index: int = 0
    name: str = "Unknown GPU"
    memory_total: int = 0  # MB
    memory_used: int = 0   # MB
    memory_free: int = 0   # MB
    utilization: int = 0   # Percentage
    temperature: int = 0   # Celsius
    power_usage: float = 0.0  # Watts
    driver_version: str = ""
    cuda_version: str = ""


class ClusterNode(BaseModel):
    """A node in the GPU cluster."""
    id: str = Field(..., description="Unique node identifier")
    name: str = Field(..., description="Display name for the node")
    host: str = Field(..., description="Hostname or IP address")
    port: int = Field(default=8888, description="Enterprise Gateway port")
    status: NodeStatus = NodeStatus.OFFLINE
    gpus: List[GPUInfo] = []
    cpu_count: int = 0
    memory_total: int = 0  # MB
    memory_available: int = 0  # MB
    active_kernels: int = 0
    max_kernels: int = 10
    last_heartbeat: Optional[datetime] = None
    created_at: datetime = Field(default_factory=datetime.utcnow)
    tags: List[str] = []  # e.g., ["training", "inference", "high-memory"]
    priority: int = 0  # Higher = preferred


class ClusterNodeCreate(BaseModel):
    """Request to add a node to the cluster."""
    name: str
    host: str
    port: int = 8888
    tags: List[str] = []
    priority: int = 0


class ClusterNodeUpdate(BaseModel):
    """Request to update a cluster node."""
    name: Optional[str] = None
    port: Optional[int] = None
    tags: Optional[List[str]] = None
    priority: Optional[int] = None
    status: Optional[NodeStatus] = None


class ClusterStats(BaseModel):
    """Overall cluster statistics."""
    total_nodes: int = 0
    online_nodes: int = 0
    total_gpus: int = 0
    available_gpus: int = 0
    total_memory: int = 0  # MB
    available_memory: int = 0  # MB
    active_kernels: int = 0
    max_kernels: int = 0


class KernelPlacement(BaseModel):
    """Request for kernel placement on a specific node."""
    node_id: Optional[str] = None  # None = auto-select best node
    gpu_index: Optional[int] = None  # None = auto-select
    require_gpu: bool = True
    min_gpu_memory: int = 0  # MB, 0 = no minimum
    tags: List[str] = []  # Required node tags

File: book/lib/models/cluster.dart
/// Cluster and GPU node models for distributed kernel execution.

enum NodeStatus {
  online,
  offline,
  busy,
  error,
  maintenance;

  String get displayName {
    switch (this) {
      case NodeStatus.online:
        return 'Online';
      case NodeStatus.offline:
        return 'Offline';
      case NodeStatus.busy:
        return 'Busy';
      case NodeStatus.error:
        return 'Error';
      case NodeStatus.maintenance:
        return 'Maintenance';
    }
  }
}

class GPUInfo {
  final int index;
  final String name;
  final int memoryTotal;
  final int memoryUsed;
  final int memoryFree;
  final int utilization;
  final int temperature;
  final double powerUsage;
  final String driverVersion;
  final String cudaVersion;

  GPUInfo({
    this.index = 0,
    this.name = 'Unknown GPU',
    this.memoryTotal = 0,
    this.memoryUsed = 0,
    this.memoryFree = 0,
    this.utilization = 0,
    this.temperature = 0,
    this.powerUsage = 0.0,
    this.driverVersion = '',
    this.cudaVersion = '',
  });

  factory GPUInfo.fromJson(Map<String, dynamic> json) {
    return GPUInfo(
      index: json['index'] ?? 0,
      name: json['name'] ?? 'Unknown GPU',
      memoryTotal: json['memory_total'] ?? 0,
      memoryUsed: json['memory_used'] ?? 0,
      memoryFree: json['memory_free'] ?? 0,
      utilization: json['utilization'] ?? 0,
      temperature: json['temperature'] ?? 0,
      powerUsage: (json['power_usage'] ?? 0.0).toDouble(),
      driverVersion: json['driver_version'] ?? '',
      cudaVersion: json['cuda_version'] ?? '',
    );
  }

  double get memoryUsagePercent => memoryTotal > 0 ? memoryUsed / memoryTotal * 100 : 0;
  String get memoryDisplay => '${(memoryUsed / 1024).toStringAsFixed(1)} / ${(memoryTotal / 1024).toStringAsFixed(1)} GB';
}

class ClusterNode {
  final String id;
  final String name;
  final String host;
  final int port;
  final NodeStatus status;
  final List<GPUInfo> gpus;
  final int cpuCount;
  final int memoryTotal;
  final int memoryAvailable;
  final int activeKernels;
  final int maxKernels;
  final DateTime? lastHeartbeat;
  final DateTime createdAt;
  final List<String> tags;
  final int priority;

  ClusterNode({
    required this.id,
    required this.name,
    required this.host,
    this.port = 8888,
    this.status = NodeStatus.offline,
    this.gpus = const [],
    this.cpuCount = 0,
    this.memoryTotal = 0,
    this.memoryAvailable = 0,
    this.activeKernels = 0,
    this.maxKernels = 10,
    this.lastHeartbeat,
    DateTime? createdAt,
    this.tags = const [],
    this.priority = 0,
  }) : createdAt = createdAt ?? DateTime.now();

  factory ClusterNode.fromJson(Map<String, dynamic> json) {
    return ClusterNode(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      host: json['host'] ?? '',
      port: json['port'] ?? 8888,
      status: _parseStatus(json['status']),
      gpus: (json['gpus'] as List<dynamic>?)
              ?.map((g) => GPUInfo.fromJson(g))
              .toList() ??
          [],
      cpuCount: json['cpu_count'] ?? 0,
      memoryTotal: json['memory_total'] ?? 0,
      memoryAvailable: json['memory_available'] ?? 0,
      activeKernels: json['active_kernels'] ?? 0,
      maxKernels: json['max_kernels'] ?? 10,
      lastHeartbeat: json['last_heartbeat'] != null
          ? DateTime.tryParse(json['last_heartbeat'])
          : null,
      createdAt: json['created_at'] != null
          ? DateTime.tryParse(json['created_at']) ?? DateTime.now()
          : DateTime.now(),
      tags: (json['tags'] as List<dynamic>?)?.cast<String>() ?? [],
      priority: json['priority'] ?? 0,
    );
  }

  static NodeStatus _parseStatus(String? status) {
    switch (status) {
      case 'online':
        return NodeStatus.online;
      case 'busy':
        return NodeStatus.busy;
      case 'error':
        return NodeStatus.error;
      case 'maintenance':
        return NodeStatus.maintenance;
      default:
        return NodeStatus.offline;
    }
  }

  int get totalGpuMemory => gpus.fold(0, (sum, gpu) => sum + gpu.memoryTotal);
  int get freeGpuMemory => gpus.fold(0, (sum, gpu) => sum + gpu.memoryFree);
  bool get isOnline => status == NodeStatus.online;
  bool get hasAvailableSlots => activeKernels < maxKernels;
}

class ClusterStats {
  final int totalNodes;
  final int onlineNodes;
  final int totalGpus;
  final int availableGpus;
  final int totalMemory;
  final int availableMemory;
  final int activeKernels;
  final int maxKernels;

  ClusterStats({
    this.totalNodes = 0,
    this.onlineNodes = 0,
    this.totalGpus = 0,
    this.availableGpus = 0,
    this.totalMemory = 0,
    this.availableMemory = 0,
    this.activeKernels = 0,
    this.maxKernels = 0,
  });

  factory ClusterStats.fromJson(Map<String, dynamic> json) {
    return ClusterStats(
      totalNodes: json['total_nodes'] ?? 0,
      onlineNodes: json['online_nodes'] ?? 0,
      totalGpus: json['total_gpus'] ?? 0,
      availableGpus: json['available_gpus'] ?? 0,
      totalMemory: json['total_memory'] ?? 0,
      availableMemory: json['available_memory'] ?? 0,
      activeKernels: json['active_kernels'] ?? 0,
      maxKernels: json['max_kernels'] ?? 0,
    );
  }

  double get nodeAvailability => totalNodes > 0 ? onlineNodes / totalNodes * 100 : 0;
  double get kernelUtilization => maxKernels > 0 ? activeKernels / maxKernels * 100 : 0;
}

class KernelPlacement {
  final String? nodeId;
  final int? gpuIndex;
  final bool requireGpu;
  final int minGpuMemory;
  final List<String> tags;

  KernelPlacement({
    this.nodeId,
    this.gpuIndex,
    this.requireGpu = true,
    this.minGpuMemory = 0,
    this.tags = const [],
  });

  Map<String, dynamic> toJson() {
    return {
      if (nodeId != null) 'node_id': nodeId,
      if (gpuIndex != null) 'gpu_index': gpuIndex,
      'require_gpu': requireGpu,
      'min_gpu_memory': minGpuMemory,
      'tags': tags,
    };
  }
}
